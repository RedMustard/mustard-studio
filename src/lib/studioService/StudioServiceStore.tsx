import { createContext, h } from 'preact';
import { useReducer } from 'preact/hooks';
import { OSC_1_INITIAL_SETTINGS, OSC_2_INITIAL_SETTINGS, OSC_SUB_INITIAL_SETTINGS } from '../../constants';
import { DispatchFunction, StudioService } from '../../types/types';
import { deepClone } from '../utils/objects/objects';
import { studioServiceReducer } from './studioServiceReducer';


const initialState: StudioService = Object.freeze({
    master: {
        gainNode: undefined,
        panNode: undefined,
        settings: {
            volume: 0.1,
            pan: 0,
        },
    },
    oscillators: {
        osc1: {
            gainNode: undefined,
            panNode: undefined,
            settings: {
                ...OSC_1_INITIAL_SETTINGS,
            },
        },
        osc2: {
            gainNode: undefined,
            panNode: undefined,
            settings: {
                ...OSC_2_INITIAL_SETTINGS,
            },
        },
        oscSub: {
            gainNode: undefined,
            panNode: undefined,
            settings: {
                ...OSC_SUB_INITIAL_SETTINGS,
            },
        },
    },
});

export const getInitialState = () => deepClone(initialState);

export const StudioServiceContext = createContext<[StudioService, DispatchFunction]>([getInitialState(), () => {}]);

export const StudioServiceStore = ({ children }: any) => {
    const [state, dispatch]: [StudioService, DispatchFunction] = useReducer(studioServiceReducer, getInitialState());

    return (
        <StudioServiceContext.Provider value={[state, dispatch]}>
            {children}
        </StudioServiceContext.Provider>
    );
};
