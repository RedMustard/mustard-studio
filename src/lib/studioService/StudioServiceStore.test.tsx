import { h } from 'preact';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import { StudioServiceStore } from './StudioServiceStore';

describe('<StudioServiceStore />', () => {
    it('renders with basic props', () => {
        const wrapper = shallow(
            <StudioServiceStore />,
        );

        expect(toJson(wrapper)).toMatchSnapshot();
    });
});
