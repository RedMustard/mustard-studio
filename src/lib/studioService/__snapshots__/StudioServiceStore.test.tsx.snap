// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<StudioServiceStore /> renders with basic props 1`] = `
<Provider
  value={
    Array [
      Object {
        "gainNodes": Object {
          "master": undefined,
          "osc1": undefined,
          "osc2": undefined,
        },
        "oscillatorNodes": Object {
          "osc1": Object {
            "enabled": true,
            "type": "sine",
          },
          "osc2": Object {
            "enabled": false,
            "type": "sawtooth",
          },
        },
        "volume": Object {
          "master": 0.1,
          "osc1": 0.5,
          "osc2": 0.5,
        },
      },
      [Function],
    ]
  }
/>
`;
